### PR TITLE
[AST] Perf: Make generic decl types share field offsets and fast casting logic

### PR DESCRIPTION
**Pull request by davezarzycki**

### CODE REVIEW

**Comment by davezarzycki** [comment]
Hi @DougGregor – Are you okay with these added abstract classes? They help improve code gen by ensuring that "TheDecl" is stored at the same offset in memory and the change makes `classof()` efficient when checking for "any generic" or "any nominal or bound generic nominal". Also, I could surely improve the comments describing these abstract classes. Is there anything you would want said or not said? Thanks!

@swift-ci please test

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -1025,6 +1025,76 @@ class alignas(1 << TypeAlignInBits) TypeBase {
...
+/// compiler itself.
+class AnyGenericType : public TypeBase {
+  friend class NominalOrBoundGenericNominalType;
+
+  /// TheDecl - This is the TypeDecl which declares the given type. It
+  /// specifies the name and other useful information about this type.
+  union {
+    GenericTypeDecl *GenDecl;
+    NominalTypeDecl *NomDecl;
+  };
```
Comment (DougGregor) [inline include/swift/AST/Types.h]: That's a clever bit of aliasing here... but it'll work fine.

**Review by DougGregor — APPROVED** [review]
LGTM. I'm fine with the extra step in the hierarchy if it improves code-gen

File: include/swift/AST/Types.h (hunk 1 above)
Comment (davezarzycki) [reply include/swift/AST/Types.h]: Thanks. Ya, the union is gross. I tried using templates, but that fought with the CanType macros. I wasn't sure what was more important. It should be polished at some point.

**Pull request merged**